### PR TITLE
External hard redirects

### PR DESCRIPTION
**Pull request by gjacobrobertson**
# Issue

Invoking `Phoenix.LiveView.redirect/2` during a `Phoenix.LiveView.mount/3` with an external URL causes an `ArgumentError` to be raised during the subsequent render.

## Example

```elixir
defmodule MyView do
  use Phoenix.LiveView

  def mount(_params, _session, socket) do
    {:ok, redirect(socket, external: "https://phoenixframework.org")}
  end
end
```

## Analysis

`Phoenix.LiveView.redirect/2`when invoked with an external URL, e.g. `redirect(socket, external: url)` causes the the `socket` to be annotated with `{:redirect, %{to: url}` in its `redirected` field, making it structurally indistiguishable from a local redirect as invoked with the `:to` option. The subsequent invocation of `Phoenix.LiveView.Controller.live_render/3` then calls `Phoenix.Controller.redirect(conn, to: to)`, which raises an `ArgumentError` if the given `to` is a fully qualified URL and not just a path.

# Proposed solution

This PR proposes a solution comprised of the following details:
- Modify `Phoenix.LiveView.redirect/2` to annotate its `socket` argument with `{:redirect, %{external: url}}` in its `redirected` field, structurally distinguishing it from a local redirect.
- Modify `Phoenix.LiveView.Controller` to pass the same key defined in the socket's `redirected` field to `Phoenix.Controller.redirect/2`, which in turn recognizes an explicitly passed external URL as valid.
- Modify `Phoenix.LiveView.Channel` to combine `:from` and `:to` annotations from the socket back into a single `:to` key for the message passed back to the client, which responds as usual by setting `window.location` accordingly.
- Update tests to reflect structural changes to socket annotations
- Provide new integration tests for invoking `Phoenix.LiveView.redirect/2` with an external URL during mounting.

### CODE REVIEW

File: lib/phoenix_live_view/channel.ex [hunk 1]
```diff
@@ -275,7 +278,11 @@ defmodule Phoenix.LiveView.Channel do
     end
   end
 
-  defp handle_result({:reply, %{} = reply, %Socket{} = new_socket}, {:handle_event, 3, ref}, state) do
+  defp handle_result(
+         {:reply, %{} = reply, %Socket{} = new_socket},
+         {:handle_event, 3, ref},
+         state
+       ) do
```
Comment (gjacobrobertson) [inline lib/phoenix_live_view/channel.ex:285]: This change is unrelated, and is the result of having run `mix format` on the file.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
